src/Donations/Migrations/RecalculateExchangeRate.php

```diff
@@ -0,0 +1,180 @@
+            ->whereIsNotNull('give_donationmeta_attach_meta_baseAmount.meta_value')
+            ->whereRaw('AND give_donationmeta_attach_meta_baseCurrency.meta_value != give_donationmeta_attach_meta_currency.meta_value')
+            ->where(function($query) {
+                $query->where('give_donationmeta_attach_meta_exchangeRate.meta_value', '1')
+                    ->orWhereIsNull('give_donationmeta_attach_meta_exchangeRate.meta_value');
+            });
+    }
+
+    /**
+     * @inheritDoc
+     * @throws DatabaseMigrationException
+     */
+    public function runBatch($firstId, $lastId): void
+    {
+        try {
+            $query = $this->query()
+                ->select('posts.ID')
+                ->orderBy('posts.ID', 'ASC');
+
+            // Migration Runner will pass null for lastId in the last step
+            if (is_null($lastId)) {
+                $query->where('posts.ID', $firstId, '>');
+            } else {
+                $query->whereBetween('posts.ID', $firstId, $lastId);
+            }
+
+            $donations = $query->getAll();
+
+            foreach ($donations as $donation) {
+                try {
+                    $amount = Money::fromDecimal($donation->amount, $donation->currency);
+                    $baseAmount = Money::fromDecimal($donation->baseAmount, $donation->baseCurrency);
+
+                    if ($baseAmount->formatToDecimal() === '0') {
+                        throw new DatabaseMigrationException(
+                            sprintf(
+                                'Invalid base amount (0) for donation ID %d',
+                                $donation->ID
+                            )
+                        );
+                    }
+
+                    /** @var Money $exchangeRate */
+                    $exchangeRate = $amount->divide($baseAmount->formatToDecimal());
+
+                    give()->payment_meta->update_meta($donation->ID, DonationMetaKeys::EXCHANGE_RATE, $exchangeRate->formatToDecimal());
+                } catch (\Exception $e) {
+                    throw new DatabaseMigrationException(
+                        sprintf(
+                            'Failed to process donation ID %d: %s',
+                            $donation->ID,
+                            $e->getMessage()
+                        ),
+                        0,
+                        $e
+                    );
+                }
+            }
+        } catch (DatabaseQueryException $exception) {
+            throw new DatabaseMigrationException(
+                'An error occurred while updating donation exchange rates',
+                0,
+                $exception
+            );
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getItemsCount(): int
+    {
+        return $this->query()->count();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getBatchItemsAfter($lastId): ?array
+    {
+        $item = clone $this->query()
```

Comment:
Also here, we don't need all that stuff from the query method; we need just the first id and the last id

Reply:
As the query method has been pretty much simplified, I don't think this will be a problem now.

1769236f0821a9a640d8de59e54bf0ca7db330fd